content/blog/2024-09-10-unlock-one-million-rps-part2.md

```diff
@@ -0,0 +1,147 @@
+
+    for (int i = 0; i < arr_size; i++) { 
+        lp = la[i]; 
+        while (lp) { 
+            res += lp->val;
+            lp = lp->next;
+        }
+    }
+
+    return res; 
+}
+``` 
+Executing this function on an array of 16 lists containing 10 million elements each takes approximately 20.8 seconds on an ARM processor (Graviton 3). Now consider the following alternative implementation which instead of scanning the lists separately,  interleaves the executions of the lists scans: 
+```c
+unsigned long interleavedSum(size_t arr_size, list **la) {
+    list **lthreads = malloc(arr_size * sizeof(list *)); 
+    unsigned long res = 0; 
+    int n = arr_size; 
+
+    for (int i = 0; i < arr_size; i++) {
+        lthreads[i] = la[i]; 
+        if (lthreads[i] == NULL) 
+            n--; 
+    } 
+
+    while(n) {
+        for (int i = 0; i < arr_size; i++) { 
+            if (lthreads[i] == NULL) 
+                continue; 
+            res += lthreads[i]->val;
+            lthreads[i] = lthreads[i]->next; 
+            if (lthreads[i] == NULL) 
+                n--;
+        }  
+    }
+
+    free(lthreads);
+    return res; 
+}
+```
+Running this new version with the same input as previously described takes less than 2 seconds, achieving a 10x speedup! The explanation for this significant improvement lies in the processor's speculative execution capabilities. In a standard sequential traversal of a linked list, as seen in the first version of the function, the processor cannot 'speculate' on future memory access instructions. This limitation becomes particularly costly with large lists, where each pointer access likely results in a expensive external memory access. In contrast, the alternative implementation, which interleaves list traversals, allows the processor to issue more memory accesses in parallel. This leads to an overall reduction in memory access latency through amortization. 
+
+One way to maximize the amount of parallel memory access issued is to add prefetch instructions. Replacing 
+```c
+             if (lthreads[i] == NULL) 
+                n--;
+```
+with
+```c 
+            if (lthreads[i]) 
+                __builtin_prefetch(lthreads[i]);
+            else 
+                n--;
+```
+reduces the execution time further to 1.8 sec. 
+
+### Back to Valkey
+
+In the first part, we described how we updated the existing I/O threads implementation to increase parallelism and reduce the amount of I/O operations executed by the main thread to a minimum. Indeed, we observed an increase in the number of requests per second, reaching up to 780K SET commands per second. Profiling the execution revealed that Valkey's main thread was spending more than 40% of its time in a single function: lookupKey, whose goal is to locate the command keys in Valkey's main dictionary. This dictionary is implemented as a straightforward chained hash, as shown in the picture below: 
+![dict find](/assets/media/pictures/lookupKey.jpg)
```

Comment:
Looks like there is case sensitivity issues. The file is .JPG but the reference is .jpg